### ISSUE TITLE
Outbound mode: endSession not closing underlying socket connection

### ISSUE THREAD

**Issue by newtondev** [davehorton/drachtio#12]
Having a slight issue with connections not being closed when running in "outbound" mode. Based off of the example code but using Srf.

```
srf.register((req, res) => {
  logger.info('srf.register socket:', res.socket);

  try {
    logger.info({ registration: req.registration }, 'Received an authenticated registration request');

    if( 'unregister' === req.registration.type ) {
      res.send(200);
    } else {
      // min_expiry
      if (req.registration.expires < 900) {
        req.registration.expires = 900;
      }

      const contactUri = req.registration.contact[0].uri;
      
      res.send(200, {
        headers: {
          'Contact': '<' + contactUri + '>;expires=' + Math.floor(response.registration.expires / 1000)
        }
      });
    }
  } catch (err) {
    res.send(503, err.message);
    // TODO: handle error accordingly!
  } finally {    
    srf.endSession(res);        
  }
});
```
The amount of client connections continues to rise on drachtio-server, and the output from the logging call in the code above for 'res.socket' is as follows:

```
[2018-04-25T09:06:31.718Z] INFO (1 on tv2-call-router-node-535718781-91n9d): srf.register socket: 
{
  "connecting": false,
  "_hadError": false,
  "_handle": {
    "bytesRead": 1610,
    "_externalStream": {
      
    },
    "fd": 194,
    "reading": true,
    "owner": "[Circular]",
    "onconnection": null,
    "writeQueueSize": 0
  },
  "_parent": null,
  "_host": null,
  "_readableState": {
    "objectMode": false,
    "highWaterMark": 16384,
    "buffer": {
      "head": null,
      "tail": null,
      "length": 0
    },
    "length": 0,
    "pipes": null,
    "pipesCount": 0,
    "flowing": true,
    "ended": false,
    "endEmitted": false,
    "reading": true,
    "sync": false,
    "needReadable": true,
    "emittedReadable": false,
    "readableListening": false,
    "resumeScheduled": false,
    "destroyed": false,
    "defaultEncoding": "utf8",
    "awaitDrain": 0,
    "readingMore": false,
    "decoder": {
      "encoding": "utf8",
      "lastNeed": 0,
      "lastTotal": 0,
      "lastChar": {
        "type": "Buffer",
        "data": [
          200,
          72,
          139,
          77
        ]
      }
    },
    "encoding": "utf8"
  },
  "readable": true,
  "domain": null,
  "_events": {
    
  },
  "_eventsCount": 7,
  "_writableState": {
    "objectMode": false,
    "highWaterMark": 16384,
    "finalCalled": false,
    "needDrain": false,
    "ending": false,
    "ended": false,
    "finished": false,
    "destroyed": false,
    "decodeStrings": false,
    "defaultEncoding": "utf8",
    "length": 0,
    "writing": false,
    "corked": 0,
    "sync": false,
    "bufferProcessing": false,
    "writecb": null,
    "writelen": 0,
    "bufferedRequest": null,
    "lastBufferedRequest": null,
    "pendingcb": 0,
    "prefinished": false,
    "errorEmitted": false,
    "bufferedRequestCount": 0,
    "corkedRequestsFree": {
      "next": null,
      "entry": null
    }
  },
  "writable": true,
  "allowHalfOpen": false,
  "_bytesDispatched": 66,
  "_sockname": {
    "address": "::ffff:10.244.200.177",
    "family": "IPv6",
    "port": 4000
  },
  "_pendingData": null,
  "_pendingEncoding": "",
  "server": {
    "domain": null,
    "_events": {
      
    },
    "_eventsCount": 2,
    "_connections": 184,
    "_handle": {
      "bytesRead": 0,
      "_externalStream": {
        
      },
      "fd": 10,
      "reading": false,
      "owner": "[Circular]",
      "onread": null,
      "writeQueueSize": 0
    },
    "_usingSlaves": false,
    "_slaves": [
      
    ],
    "_unref": false,
    "allowHalfOpen": false,
    "pauseOnConnect": false,
    "_connectionKey": "6::::4000"
  },
  "_server": {
    "domain": null,
    "_events": {
      
    },
    "_eventsCount": 2,
    "_connections": 184,
    "_handle": {
      "bytesRead": 0,
      "_externalStream": {
        
      },
      "fd": 10,
      "reading": false,
      "owner": "[Circular]",
      "onread": null,
      "writeQueueSize": 0
    },
    "_usingSlaves": false,
    "_slaves": [
      
    ],
    "_unref": false,
    "allowHalfOpen": false,
    "pauseOnConnect": false,
    "_connectionKey": "6::::4000"
  },
  "_peername": {
    "address": "::ffff:10.155.250.70",
    "family": "IPv6",
    "port": 57810
  },
  "_consuming": true
}
```

The "_connections": 184, never reduces but just keeps climbing even after calling srf.endSession(res). It gets to the point where the file descriptors are exhausted and we need to restart the application to get rid of the connections.

Using npm packages:
"drachtio": "^4.1.9",
"drachtio-mw-registration-parser": "0.0.2",
"drachtio-srf": "^4.2.11",
"passport": "^0.4.0",
"passport-http": "^0.3.0"

Authentication is using passport middleware with DigestAuthentication.

Please let me know if you require any additional information. Thanks

**Comment by davehorton** [comment]
I'm trying to recreate this but getting different results -- connections are *not* growing.  My test program is below, and I am using `drachtio-srf@4.2.11`, can you let me know what version you are running so I can test with that?

I created a similar test program to yours, and then set a softphone client to register every 20 seconds.  (Note: by running with DEBUG=drachtio:socket you can see more detail on socket connections):

```
DEBUG=drachtio:socket node app.js --port 3002 --secret cymru
  drachtio:socket _initServer: added socket: 55578:3002, count now: 1 +0ms
  drachtio:socket _initServer: added socket: 55579:3002, count now: 2 +711ms
  drachtio:socket disconnect: removing socket 55579:3002 +12ms
  drachtio:socket disconnect: after delete there are 1 entries in mapServer +0ms
after processing register res.socket._server._connections: 2
  drachtio:socket _initServer: added socket: 55580:3002, count now: 2 +853ms
  drachtio:socket disconnect: removing socket 55580:3002 +493ms
  drachtio:socket disconnect: after delete there are 1 entries in mapServer +0ms
after processing unregister res.socket._server._connections: 2
  drachtio:socket _initServer: added socket: 55582:3002, count now: 2 +568ms
  drachtio:socket disconnect: removing socket 55582:3002 +13ms
  drachtio:socket disconnect: after delete there are 1 entries in mapServer +0ms
after processing register res.socket._server._connections: 2
  drachtio:socket _initServer: added socket: 55603:3002, count now: 2 +16s
  drachtio:socket disconnect: removing socket 55603:3002 +11ms
  drachtio:socket disconnect: after delete there are 1 entries in mapServer +0ms
after processing register res.socket._server._connections: 2
  drachtio:socket _initServer: added socket: 55605:3002, count now: 2 +15s
  drachtio:socket disconnect: removing socket 55605:3002 +11ms
  drachtio:socket disconnect: after delete there are 1 entries in mapServer +0ms
after processing register res.socket._server._connections: 2
  drachtio:socket _initServer: added socket: 55606:3002, count now: 2 +15s
  drachtio:socket disconnect: removing socket 55606:3002 +11ms
  drachtio:socket disconnect: after delete there are 1 entries in mapServer +1ms
after processing register res.socket._server._connections: 2
  drachtio:socket _initServer: added socket: 55613:3002, count now: 2 +15s
  drachtio:socket disconnect: removing socket 55613:3002 +10ms
  drachtio:socket disconnect: after delete there are 1 entries in mapServer +0ms
after processing register res.socket._server._connections: 2
  drachtio:socket _initServer: added socket: 55614:3002, count now: 2 +15s
  drachtio:socket disconnect: removing socket 55614:3002 +11ms
  drachtio:socket disconnect: after delete there are 1 entries in mapServer +0ms
after processing register res.socket._server._connections: 2
  drachtio:socket _initServer: added socket: 55619:3002, count now: 2 +15s
  drachtio:socket disconnect: removing socket 55619:3002 +11ms
  drachtio:socket disconnect: after delete there are 1 entries in mapServer +1ms
```
Here is my test program in full:
```
const Srf = require('drachtio-srf') ;
const srf = new Srf();
const passport       = require('passport') ;
const DigestStrategy = require('passport-http').DigestStrategy;
const registrationParser = require('drachtio-mw-registration-parser') ;
const argv = require('minimist')(process.argv.slice(2));
const realm = 'sip.drachtio.org';

srf.listen({
  port: argv.port,
  secret: argv.secret
}) ;
const users = [
  { id: 1, username: 'dhorton', password: 'pass1234', domain: realm}
];
function findByUsername(username, fn)
{
  for (let i = 0, len = users.length; i < len; i++) {
    const user = users[i];
    if (user.username === username) { return fn(null, user); }
  }
  return fn(null, null);
}

passport.use(
  new DigestStrategy({qop: 'auth', realm: realm}, (username, done) => {
    // Find the user by username. If there is no user with the given username
    // set the user to `false` to indicate failure. Otherwise, return the
    // user and user's password.

    findByUsername(username, (err, user) => {
      if (err) return done(err);
      if (!user) return done(null, false);
      return done(null, user, user.password);
    }
    );
  }));

srf.use(passport.initialize());
srf.use('register', registrationParser);
srf.use('register', passport.authenticate('digest', {session: false})) ;

srf.register((req, res) => {
  if ('unregister' === req.registration.type) {
    res.send(200) ;
  }
  else {
    res.send(200, {
      headers: {
        'Contact': `<${req.registration.aor}>;expires=20`
      }
    }) ;
  }
  srf.endSession(res);
  console.log(`after processing ${req.registration.type} res.socket._server._connections: ${res.socket._server._connections}`);
});
```

**Comment by davehorton** [comment]
I see you provided the versions in your initial post; sorry I missed that.

Hmm, it looks like you are using the same version of drachtio-srf as I am, but getting different results.  Note that drachtio itself is no longer an explicit dependency (it will be brought in by drachtio-srf).

Here are my dependencies in full:
```
  "dependencies": {
    "drachtio-mw-registration-parser": "0.0.2",
    "drachtio-srf": "^4.2.11",
    "minimist": "^1.2.0",
    "passport": "^0.4.0",
    "passport-http": "^0.3.0"
  }
```

**Comment by davehorton** [comment]
Can you run for a bit with DEBUG=drachtio:socket and provide the output of that?  Maybe modify your logging statement to just output `res.socket._server._connections` instead of the full object to be more concise

**Comment by newtondev** [comment]
Hi Dave, thanks for the replies. Will do what you advise and provide you with the results shortly.

**Comment by newtondev** [comment]
Hi Dave,

Below is some output, I sent through one registration.

[2018-04-25T14:34:13.255Z] INFO (1 on tv2-call-router-node-2181728684-sdph2): Ending session for SRF (1.3)
Wed, 25 Apr 2018 14:34:13 GMT drachtio:socket disconnect: removing socket 41800:4000
Wed, 25 Apr 2018 14:34:13 GMT drachtio:socket disconnect: after delete there are 2 entries in mapServer
after processing register res.socket._server._connections: 3
Wed, 25 Apr 2018 14:34:45 GMT drachtio:socket _initServer: added socket: 41804:4000, count now: 3
Wed, 25 Apr 2018 14:34:53 GMT drachtio:socket _initServer: added socket: 41808:4000, count now: 4
Wed, 25 Apr 2018 14:35:40 GMT drachtio:socket _initServer: added socket: 41812:4000, count now: 5
Wed, 25 Apr 2018 14:35:48 GMT drachtio:socket _initServer: added socket: 41816:4000, count now: 6
Wed, 25 Apr 2018 14:36:35 GMT drachtio:socket _initServer: added socket: 41820:4000, count now: 7
Wed, 25 Apr 2018 14:36:44 GMT drachtio:socket _initServer: added socket: 41824:4000, count now: 8
Wed, 25 Apr 2018 14:37:30 GMT drachtio:socket _initServer: added socket: 41830:4000, count now: 9
Wed, 25 Apr 2018 14:37:39 GMT drachtio:socket _initServer: added socket: 41834:4000, count now: 10
Wed, 25 Apr 2018 14:38:25 GMT drachtio:socket _initServer: added socket: 41838:4000, count now: 11
Wed, 25 Apr 2018 14:38:34 GMT drachtio:socket _initServer: added socket: 41842:4000, count now: 12
Wed, 25 Apr 2018 14:39:20 GMT drachtio:socket _initServer: added socket: 41846:4000, count now: 13
Wed, 25 Apr 2018 14:39:29 GMT drachtio:socket _initServer: added socket: 41850:4000, count now: 14
Wed, 25 Apr 2018 14:40:15 GMT drachtio:socket _initServer: added socket: 41854:4000, count now: 15
Wed, 25 Apr 2018 14:40:25 GMT drachtio:socket _initServer: added socket: 41858:4000, count now: 16
Wed, 25 Apr 2018 14:41:10 GMT drachtio:socket _initServer: added socket: 41862:4000, count now: 17
Wed, 25 Apr 2018 14:41:20 GMT drachtio:socket _initServer: added socket: 41866:4000, count now: 18

I think I managed to see what is increasing the connection count, log on the drachtio-server shows OPTIONS requests going through which are not handled by my application, so no srf.endSession call to close those.

2018-04-25 14:45:01.660454 Client::processMessage - got request with 4 tokens
2018-04-25 14:45:01.660525 Client::processMessage - request id f4dc13b3-14ec-4b81-b760-3a14b596cf60, request type: sip transaction id: c7d14a01-fba7-48bc-bcdc-22bd679ca149, dialog id: 
2018-04-25 14:45:01.660577 ClientController::addApiRequest: clientMsgId f4dc13b3-14ec-4b81-b760-3a14b596cf60; size: 1
2018-04-25 14:45:01.660718 SipDialogController::doRespondToSipRequest thread 7f60bbd88780
2018-04-25 14:45:01.660807 SipDialogController::findAndRemoveTransactionIdForIncomingRequest - searching transactionId c7d14a01-fba7-48bc-bcdc-22bd679ca149
2018-04-25 14:45:01.660882 SipDialogController::findAndRemoveTransactionIdForIncomingRequest - failed to find transactionId c7d14a01-fba7-48bc-bcdc-22bd679ca149
2018-04-25 14:45:01.660984 pending-request: removing entry, prior to removal length: 1
2018-04-25 14:45:01.661058 pending-request: removed entry, timer not set (queue is empty after removal), length: 0
2018-04-25 14:45:01.661151 SipDialogController::addIncomingRequestTransaction - adding transactionId c7d14a01-fba7-48bc-bcdc-22bd679ca149 for irq:0x157a730
2018-04-25 14:45:01.661229 SipDialogController::findAndRemoveTransactionIdForIncomingRequest - searching transactionId c7d14a01-fba7-48bc-bcdc-22bd679ca149
2018-04-25 14:45:01.661296 SipDialogController::doRespondToSipRequest found incoming transaction 0x157a730
2018-04-25 14:45:01.661362 SipTransport::getContactUri - created Contact header: sip:197.155.250.156:6060
2018-04-25 14:45:01.661549 makeTags - Adding well-known header 'Call-ID' with value '1974806854@192.168.1.51'
2018-04-25 14:45:01.661610 makeTags - Adding well-known header 'cseq' with value '1 OPTIONS'
2018-04-25 14:45:01.661659 makeTags - Adding well-known header 'from' with value '"David Meintjes" <sip:5a8eb1299c4d5e0011260839@sip1.fibrephone.telviva.com:6060>;tag=450987727'
2018-04-25 14:45:01.661713 makeTags - Adding well-known header 'to' with value '<sip:sip1.fibrephone.telviva.com:6060>'
2018-04-25 14:45:01.661816 tport.c:3285 tport_tsend() tport_tsend(0xa92480) tpn = TCP/197.95.208.152:37797
2018-04-25 14:45:01.661931 tport.c:3531 tport_send_msg() tport_vsend returned 334
2018-04-25 14:45:01.662100 send 334 bytes to tcp/[197.95.208.152]:37797 at 14:45:01.661877:
SIP/2.0 404 Not Found
Via: SIP/2.0/TCP 197.95.208.152:5162;branch=z9hG4bK2641146474;rport=37797
From: "David Meintjes" <sip:5a8eb1299c4d5e0011260839@sip1.fibrephone.telviva.com:6060>;tag=450987727
To: <sip:sip1.fibrephone.telviva.com:6060>;tag=U8g3Da91p8F8e
Call-ID: 1974806854@192.168.1.51
CSeq: 1 OPTIONS
Content-Length: 0
 
2018-04-25 14:45:01.662177 tport.c:2324 tport_set_secondary_timer() tport(0xa92480): reset timer
2018-04-25 14:45:01.662234 nta.c:6879 incoming_reply() nta: sent 404 Not Found for OPTIONS (1)
2018-04-25 14:45:01.662281 SipDialogController::doRespondToSipRequest destroying irq 0x157a730
2018-04-25 14:45:01.662349 ClientController::removeApiRequest: clientMsgId f4dc13b3-14ec-4b81-b760-3a14b596cf60; size: 0
2018-04-25 14:45:01.662425 ClientController::removeNetTransaction: transactionId c7d14a01-fba7-48bc-bcdc-22bd679ca149; size: 1
2018-04-25 14:45:01.662500 nta.c:5832 incoming_free() nta: incoming_free(0x157a730)
2018-04-25 14:45:01.664582 Client::write_handler - wrote 529 bytes: system:0

**Comment by davehorton** [comment]
ah, I see.  Yes, the library itself should clear those.  I will work on recreating and then providing a fix.

**Comment by davehorton** [comment]
in the meantime you can add a simple `srf.options` handler, send a 200 OK and then call endSession

**Comment by newtondev** [comment]
Hi Dave,

Thanks so much. Added the code below and now the connections are no longer growing in size.

```
if ('outbound' === argv.type) {
  srf.options((req, res) => {
    res.send(200);
      
    srf.endSession(res);
    console.log(`after processing srf.options res.socket._server._connections: ${res.socket._server._connections}`);
  });
}
```

**Comment by davehorton** [comment]
great.  FYI you can probably eliminate the check about the outbound type since the library will do that for you and silently ignore the 'endSession' call in that case, so just 
```
srf.options((req, res) => {
    res.send(200);
      
    srf.endSession(res);
    console.log(`after processing srf.options res.socket._server._connections: ${res.socket._server._connections}`);
  });
```
should work equivalently

**Comment by davehorton** [comment]
fixed in b147dc5ede1d8c7496ccdb474ad203be6eba6839

available in drachtio-srf@4.2.12
